### ISSUE TITLE
Options to interpret missing metrics

### ISSUE THREAD

**Issue by jbusecke** [xgcm/xgcm#164]
Working with xgcm data has made me acutely aware that for user friendlyness, we should probably have a keyword argument to infer missing grid metrics. 

It is possible to infer some variables from 'CF bounds' (#127), but that will usually only give you either 1) the dx/dy centered at the tracer point - if `lat_bounds` and `lon_bounds` are given as two element arrays or 2) the dx at the north face and dy at the eastface (if the bounds are given as 4 element array; usually called `vertex` or similar). 

In either case, many distances are missing. I think it would be a reasonable default if these would be interpolated from surrounding metrics. The danger here is that people would never pass the exact metrics, and perhaps this should not be turned on by default, but I think it needs to be available.

This should be considered in conjunction with #143, since I think it is important the resulting data carries metadata that clearly states which distances were 'inferred' and might thus be less accurate.

**Comment by jbusecke** [comment]
@dcherian @ALDepp thoughts on this?

**Comment by miniufo** [comment]
First of all, **interpret** in the title should be **interpolate** missing metrics?

Up to now, I think I've encountered this problem many times and the error prompt is not clear enough for me to reach a solution.   Here is one example #185, and another is when I tried to perform vertical derivative along `Z` using standard MITgcm outputs (simulated in x-z plane) as `grid.derivative(var, 'Z', boundary='fill')`, error of `cannot find any combination of...` shows up again.  Standard MITgcm outputs only have `drC` and `drF` as vertical metrics.  I followed the horizontal case and added a metric `drG = xr.DataArray(dset.drC[:-1].values, dims='Zl', coords={'Zl':dset.Zl.values})` to the `grid` object and the error just gone.  But I'm not sure if it is correct.

So I just wonder if there is a **complete** list (or doc) of the metrics needed for all core functions in `xgcm` like `average`, `integrate`, `diff`...  Users are very unlikely to know exactly what metrics are needed when calling these functions (maybe you guys know it well).  I read the [doc here](https://xgcm.readthedocs.io/en/latest/grid_metrics.html) to get the area metrics in the horizontal plane.  How about the area metrics in the x-z plane if my simulation is in this case?  Do I need something like:
```python
metrics = {
    ('X',): ['dxC', 'dxG'], # X distances
    ('Z',): ['drW', 'drS', 'drC'], # Z distances
    ('X', 'Z'): [...some area metrics here...] # Areas
}
```
For models other than MITgcm, they may use other names of cell boundaries (e.g., `xu_ocean` in `mom` rather than `XG` in `MITgcm`).  Is it possible to defined a unified nomenclature so that the task of explicitly specifying metrics will be passed to `autogenerate` module with a kwarg `model` like `grid = Grid(dset, model='mitgcm', infer_full_metrics=True)`?  I have to admit that I'm a lazy guy ∩_∩.

**Comment by jbusecke** [comment]
> I have to admit that I'm a lazy guy

Me too, thats why I like xgcm so much haha. 

But jokes aside, I think you are absolutely right with your concern. I think a default option to interpolate metrics makes sense, **BUT** you actually make a good point: The user needs to understand the grid of their model! 
This step will not dissapear with any defaults, and I think we are not aiming for xgcm to make it a package that allows the user to fly blind, so to speak.
 To a certain degree you will always have to have some knowledge about your specific setup to reach the highest possible accuracy. In many cases the model output just does not have all the metrics that would be needed. In your example you seem to be trying to be missing `rf` [see here](http://mitgcm.org/public/r2_manual/latest/online_documents/node48.html). For mitgcm the situation is usually pretty good since they have a nice manual, but for other models these infos are not as easy to come by. 

That said, I have been talking to @rabernat about refactoring the `autogenerate` module in the future and enabling it to be called as an additional option in `Grid(...)`, but this will take some time. I think an intermediate solution would be to add a more detailed discussion to the docs (also requested in #194 and elsewhere). 

Would you be willing to come up with an example how to interpolate metrics with the example dataset we have? You could even take some metrics away on purpose and then 'recreate' them to show where these differences matter or not. This would be a very welcome addition to the project and a great first PR for you.

**Comment by miniufo** [comment]
I have sensed the power of `xgcm`.  A lot of enthusiasms here and I'll make them count.  So no template?  I could also start from the example in the `doc` directory.  It may take a few days to write some notebooks on what I have right now.

**Comment by jbusecke** [comment]
I think these are two different things. 1) The template to show how to work with several models (which I am shamefully overdue to create)
2) Additional text in the actual docs about the [metric feature](https://xgcm.readthedocs.io/en/latest/grid_metrics.html). 
I imagined a little warning box that explains how the metrics are constructed (e.g. give all metrics containing an x distance with the key 'X' and all metrics describing an area in the x-y-plane with 'XY') + a very short line about how to interpolate metrics and then use them?

**Comment by miniufo** [comment]
I've sent a PR #196  that add a jupyter notebook on how to infer missing metrics in several cases, in the doc directory.  Please tell me if it can be improved in any way (not sure if you can preview it before merge.).

**Comment by jbusecke** [comment]
This will be closed by #345
